src/stake_table/mod.rs

```diff
@@ -1,198 +1,268 @@
 use self::{
     error::StakeTableError,
-    utils::{to_merkle_path, PersistentMerkleNode},
-};
-use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
-use ark_std::{
-    collections::HashMap,
-    rand::{CryptoRng, RngCore},
-    sync::Arc,
-    vec::Vec,
+    utils::{to_merkle_path, Key, PersistentMerkleNode},
 };
+use ark_std::{collections::HashMap, rand::SeedableRng, sync::Arc, vec::Vec};
+use digest::crypto_common::rand_core::CryptoRngCore;
 use ethereum_types::{U256, U512};
 use serde::{Deserialize, Serialize};
-use tagged_base64::tagged;
 
 mod config;
 mod utils;
 
 // Exports
 pub mod error;
-pub use utils::MerkleCommitment;
-pub use utils::MerklePath;
-pub use utils::MerklePathEntry;
-pub use utils::MerkleProof;
-
-/// Copied from HotShot repo.
-/// Type saftey wrapper for byte encoded keys.
-/// Assume that the content is a canonically serialized public key
-#[tagged("PUBKEY")]
-#[derive(
-    Clone, Debug, Hash, CanonicalSerialize, CanonicalDeserialize, PartialEq, Eq, PartialOrd, Ord,
-)]
-pub struct EncodedPublicKey(pub Vec<u8>);
-
-/// Enum type for stake table version
-///  * `STVersion::PENDING`: the most up-to-date stake table, where the incoming transactions shall be performed on.
-///  * `STVersion::FROZEN`: when an epoch ends, the PENDING stake table is frozen for leader elections for next epoch.
-///  * `STVersion::ACTIVE`: the active stake table for leader election.
-pub enum STVersion {
-    PENDING,
-    FROZEN,
-    ACTIVE,
+pub use utils::{MerkleCommitment, MerklePath, MerklePathEntry, MerkleProof};
+
+/// Snapshots of the stake table
+/// - the latest "Head" where all new changes are applied to
+/// - `EpochStart` marks the snapshot at the beginning of the current epoch
+/// - `LastEpochStart` marks the beginning of the last epoch
+/// - `BlockNum(u64)` at arbitrary block height
+pub enum SnapshotVersion {
+    Head,
+    EpochStart,
+    LastEpochStart,
+    BlockNum(u64),
+}
+
+/// Common interfaces required for a stake table used in HotShot System.
+/// APIs that doesn't take `version: SnapshotVersion` as an input by default works on the head/latest version.
+pub trait StakeTableScheme {
+    /// type for stake key
+    type Key: Clone;
+    /// type for the staked amount
+    type Amount: Clone + Copy;
+    /// type for the commitment to the current stake table
+    type Commitment;
+    /// type for the proof associated with the lookup result (if any)
+    type LookupProof;
+    /// type for the iterator over (key, value) entries
+    type IntoIter: Iterator<Item = (Self::Key, Self::Amount)>;
+
+    /// Register a new key into the stake table.
+    fn register(&mut self, new_key: Self::Key, amount: Self::Amount)
```

Comment:
Oh you remind me, since `Key` only requires `Clone`, should we use `&Self::Key` here and make a clone inside? Or we could also require `Copy`.